contracts/StoryProtocolGateway.sol

```diff
@@ -1,503 +1,424 @@
+        SignatureData calldata sigMetadata,
+        SignatureData calldata sigRegister
+    ) external returns (address ipId) {
+        ipId = IP_ASSET_REGISTRY.register(block.chainid, nftContract, tokenId);
+        _setMetadataWithSig(metadata, ipId, sigMetadata);
+        _setPermissionForModule(
+            ipId,
+            sigRegister,
+            address(LICENSING_MODULE),
+            ILicensingModule.registerDerivativeWithLicenseTokens.selector
         );
+        LICENSING_MODULE.registerDerivativeWithLicenseTokens(ipId, licenseTokenIds, royaltyContext);
+    }
 
-        ipId = IP_ASSET_REGISTRY.register(
-            block.chainid,
-            tokenContract,
-            tokenId,
-            address(metadataResolver),
-            true,
-            canonicalMetadata
-        );
+    /// @dev Registers PIL License Terms and attaches them to the given IP.
+    /// @param ipId The ID of the IP.
+    /// @param terms The PIL terms to be registered.
+    /// @return licenseTermsId The ID of the registered PIL terms.
+    function _registerPILTermsAndAttach(
+        address ipId,
+        PILTerms calldata terms
+    ) internal returns (uint256 licenseTermsId) {
+        licenseTermsId = PIL_TEMPLATE.registerLicenseTerms(terms);
+        LICENSING_MODULE.attachLicenseTerms(ipId, address(PIL_TEMPLATE), licenseTermsId);
     }
 
-    /// @dev Sets permission to allow SPG call other modules on behalf of user with user's signature.
-    /// @param ipId The address of the IP asset to set the permission for.
-    /// @param signer The address of the signer to set the permission for.
-    /// @param deadline The deadline for the signature to be valid.
-    /// @param signature The signature to set the permission for the IP.
-    function _setPermissionWithSig(address ipId, address signer, uint256 deadline, bytes calldata signature) internal {
-        AccessPermission.Permission[] memory permissionList = new AccessPermission.Permission[](2);
-        permissionList[0] = AccessPermission.Permission({
-            ipAccount: ipId,
-            signer: address(this),
-            to: address(LICENSING_MODULE),
-            func: bytes4(0),
-            permission: AccessPermission.ALLOW
-        });
-        permissionList[1] = AccessPermission.Permission({
-            ipAccount: ipId,
-            signer: address(this),
-            to: address(metadataResolver),
-            func: bytes4(0),
-            permission: AccessPermission.ALLOW
-        });
+    /// @dev Transfers the license tokens from the caller to this contract.
+    /// @param licenseTokenIds The IDs of the license tokens to be transferred.
+    function _transferLicenseTokens(uint256[] calldata licenseTokenIds) internal {
+        if (licenseTokenIds.length == 0) revert Errors.SPG__EmptyLicenseTokens();
+        for (uint256 i = 0; i < licenseTokenIds.length; i++) {
+            LICENSE_TOKEN.transferFrom(msg.sender, address(this), licenseTokenIds[i]);
+        }
```

Comment:
Current logic requires the owner of LicenseToken have to approve SPG contract to transfer LT on behalf of the LT owners.
For the better UX, we can check the ownership fo the LTs first, if the owner has already transfer the LT to SPG, then we don't need to try to transfer the LT again.

Reply:
Resolved with 1ad6121